Common/SensorFusion/Inc/SensorFusion.hpp

```diff
@@ -0,0 +1,138 @@
+} SFError_t;
+
+typedef struct {
+    float roll, pitch, yaw; //Degrees. Yaw of 180 is north.
+    float rollRate, pitchRate, yawRate; //Degrees/second
+    float airspeed; //m/s
+    float altitude; //m
+    float rateOfClimb; //m/s
+    long double latitude; //Decimal degrees
+    float latitudeSpeed; //m/s
+    long double longitude; //Decimal degrees
+    float longitudeSpeed; //m/s
+    double track; //Degrees. Track of 0 is north.
+    float groundSpeed; //m/s
+    double heading; //Degrees. Heading of 0 is north.
+} SFOutput_t;
+
+//Following structs store the raw sensor data so other modules can have direct access to them without including sensor header files
+typedef struct
+{
+    float magx, magy, magz;
+    float accx, accy, accz;
+    float gyrx, gyry, gyrz; 
+    float temp;
+    
+    bool isDataNew; 
+    enum SensorErrorCodes sensorStatus; 
+    float utcTime; 
+} IMU_Data_t;
+
+typedef struct 
+{
+    double airspeed;        
+
+    enum SensorErrorCodes sensorStatus;       
+    bool isDataNew;         
+    float utcTime;          
+} Airspeed_Data_t;
+
+typedef struct
+{
+    long double latitude;  // 8 Bytes
+    long double longitude; // 8 Bytes
+    float utcTime;     // 4 Bytes. Time in seconds since 00:00 (midnight)
+    float groundSpeed; // in m/s
+    int altitude; // in m
+    int16_t heading; // in degrees. Should be between 0-360 at all times, but using integer just in case
+    uint8_t numSatellites;    // 1 Byte
+    uint8_t fixStatus; //0 = No GPS, 1 = GPS fix, 2 = DGSP Fix, 3 = Estimated/Dead Recoking Fix
+
+    enum SensorErrorCodes sensorStatus; // 0 = no fix, 1 = gps fix, 2 = differential gps fix (DGPS) (other codes are possible)
+    bool dataIsNew; // true if data has been refreshed since the previous time GetResult was called, false otherwise.
+	bool timeIsNew;
+
+    //Added these so autopilot knows which data is new
+    bool ggaDataIsNew; //Position, altitude, time, and number of satellites
+    bool vtgDataIsNew; //Groundspeed and Heading
+}  Gps_Data_t;
+
+typedef struct  {
+
+    float pressure, altitude, temp;
+
+    bool isDataNew; 
+    enum SensorErrorCodes status; //TBD but probably 0 = SUCCESS, -1 = FAIL, 1 = BUSY 
+    int utcTime; //Last time GetResult was called
+} Altimeter_Data_t;
+
+
+/**
+ * Initialize sensor fusion.
+ */ 
+void SF_Init(void);
+
+/**
+ * Generates fused sensor data. Should be called at a constant rate defined by SF_FREQ after SF_Init has been called once.
+ */ 
+SFError_t SF_GenerateNewResult();
+
+/**
+ * Get latest fused sensor data. Can be called any time data is needed after SF_init has been called once. Waits until the output struct is not being accessed by another task.
+ * @param [out] output Output struct for fused data.
+ * @return Error struct.
+ */ 
+SFError_t SF_GetResult(SFOutput_t *SFoutput);
+
+/**
+ * Get raw IMU data. Can be called any time raw data is needed.
+ * @return IMU struct.
+ */ 
+IMU_Data_t SF_GetRawIMU();
+
+#ifdef AUTOPILOT
```

Comment:
has the distinction between AUTOPILOT and SAFETY been made through CMake or other means?

Reply:
No, but by default AUTOPILOT will be undefined so it should compile for safety. When we want to delete the old sensor fusion code from autopilot we can replace it with this version by defining AUTOPILOT in CMake.

Reply:
Sounds good